docs/upgrading/migrating-local-extensions-33.md

```diff
@@ -0,0 +1,138 @@
+---
+title: "Migrating Local Extensions for Bolt 3.3"
+---
+Migrating Local Extensions for Bolt 3.3
+===========================================
+
+Bolt 3.3 brings with it a fairly big refactor of local extensions and
+unavoidably this will mean some breaking changes.
+
+The good news is that by changing the approach a little we will hopefully make
+it a lot easier to write and add local extensions to your applications.
+
+## The Essential Migration Steps
+
+### Autoloading
+
+Local extensions are not autoloaded within the extensions directory anymore so
+you will need to add each local extensions directory to the `composer.json`
+file in `extensions/`.
+
+As an example, if you currently have a local extension at:
+`extensions/local/myname/myextension` your extension class may either be in that
+directory, or perhaps inside a `src` directory.
+
+Let's assume the class is in the root directory, you need to open your root composer.json
+file and add this to the autoload/psr-4 section.
+
+    "Bolt\\Extension\\Myname\\MyExtensionNamespace\\": "local/myname/myextension"
+    
+If you don't currently have a psr-4 section in your `extensions/composer.json` file, then
+add it so it looks something like this
+
+```json
+    "autoload" : {
+        "psr-4" : {
+            "Bolt\\Extension\\Myname\\MyExtensionNamespace\\": "local/myname/myextension"
+        }
+    }
+```
+
+You need to make one entry for each of your local extensions. The setup here is
+identical to how you would setup your own application and there's absolutely no
+need to keep your extensions inside `extensions/local` any more. You may well
+decide to move extensions to be part of your main App namespace and if so you
+can do something like the following: 
+
+```json
+    "autoload" : {
+        "psr-4" : {
+            "MyCompany\\MyApp\\": "src"
+        }
+    }
+```
```

Comment:
> Personally I'd prefer pointing to the main composer.json since it keeps the extensions one more of a sandbox for us to adjust as needed.

I agree 100%

> @bobdenotter suggested it may be less of a leap to use the extensions one.

I think it is a necessary leap, and a very small one at that. There's no reason to half way use `extensions/composer.json` and half way not. That causes confusion for both users and us devs.

Keep `extensions/*` a black box for those that want to just install extensions from the UI, that's fine. But anyone loading / creating extensions themselves can use the main `composer.json` for autoloading and add the extension to the bootstrap list themselves. This also prevents NIV here syndrome, as autoloading is the same everywhere.